proof/encoding.go

```diff
@@ -499,3 +501,141 @@ func DUint32Option(r io.Reader, val interface{}, buf *[8]byte, l uint64) error {
 	}
 	return tlv.NewTypeForDecodingErr(val, "*fn.Option[uint32]", l, l)
 }
+
+// UrlSliceOptionEncoder encodes a URL option. If the value is not set, or the
+// slice is empty, we'll encode it as a zero-length record. But that means that
+// the type and length fields will still be encoded, which is different from the
+// record not being present at all. If the distinction should be made (e.g. to
+// not re-encode old records that didn't have that field at all with the new
+// zero-length record), the caller needs to handle that by conditionally
+// including or not including the record.
+func UrlSliceOptionEncoder(w io.Writer, val any, buf *[8]byte) error {
```

Comment:
```
If the value is not set, or the
// slice is empty, we'll encode it as a zero-length record.
```

Why are we doing it that way? That's not what I would expect from encoding an `Option`. If it's `None` then the record shouldn't be included.

What's the point in encoding a `None` with as a zero length record?

Reply:
The thing is, once you're in the `...Encoder` function, you can't decide to _not_ include a record anymore. That's why we only conditionally add records in `EncodeRecords()`. That's what I mean with the part in the comment that says `the caller needs to handle that by conditionally including or not including the record.`.

All of this goes away once we use the new TLV encoding way, when this would become a `tlv.OptionalRecordT`.

Reply:
Why can't we not encode within the `...Encoder`?

Does this have strange implications for reproducing old proofs in the decode/encode cycle?

Reply:
We can decide to not encode any _payload_ within the `...Encoder` function. But when this function is called, then the T and L of TLV are already written to the stream (assuming we didn't skip the record altogether as we do in https://github.com/lightninglabs/taproot-assets/blob/6c68c9f81fec2dd5967d454918a7c904a74cf652/proof/meta.go#L492). So once we are in the encoder function, we cannot decide whether to include the record or not. We can only decide whether we write any data or cause it to be a zero-length record (but the type and length would still be written).

Reply:
The other way I think about this is why not just have `UrlSliceEncoder` and decoder and then relectively call based on some/none. So that the encoder/decoder functionality (and its doc comment) doesn't have to express the None case complexity.